File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    dic ={}
    for i in range(len(plays)):
        if genres[i] in dic:
            dic[genres[i]]['total'] += plays[i]
            dic[genres[i]]['plays'].append((plays[i],i))
        else: dic[genres[i]]={'total': plays[i], 'plays':[(plays[i], i)]}
        
    for key in dic:
        dic[key]['plays'] = sorted(dic[key]['plays'], key=lambda x: x[0], reverse=True)
    
    dic = sorted(dic.items(), key=lambda x: x[1]['total'], reverse=True)    # x[0] = key, x[1] = value

    for i in range(len(dic)):
        for j in range(len(dic[i][1]['plays'])):
            if j<2:
                answer.append(dic[i][1]['plays'][j][1])
            else: break
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    const dict = {};
    let gp = genres.map((v,i)=>[plays[i],v, i]);
    
    gp.sort((a,b)=>b[0]-a[0]);
    gp.map(([play,genre,idx])=>{
        if(dict[genre]) {
            dict[genre][0]+=play;
            if (dict[genre][1].length <2) dict[genre][1].push(idx);
        }
        else dict[genre]=[play,[idx]];
    })
    
    gp = Object.values(dict);
    gp.sort((a,b)=>b[0]-a[0]);
    gp.forEach(v=>answer.push(...v[1]));
    return answer;
}